module/execution/registers_async.go

```diff
@@ -0,0 +1,65 @@
+package execution
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/state_synchronization"
+	"github.com/onflow/flow-go/storage"
+	"go.uber.org/atomic"
+)
+
+// RegistersAsyncStore has the same basic structure as access/backend.ScriptExecutor
+// TODO: use this implementation in the `scripts.ScriptExecutor` passed into the AccessAPI
+type RegistersAsyncStore struct {
+	registerIndex storage.RegisterIndex
+	reporter      state_synchronization.IndexReporter
+	initialized   *atomic.Bool
+	init          sync.Once
+}
+
+func NewRegistersAsyncStore() *RegistersAsyncStore {
+	return &RegistersAsyncStore{
+		initialized: atomic.NewBool(false),
+	}
+}
+
+// InitDataAvailable follows the same pattern of backend.ScriptExecutor
+// This method can be called at any time after the RegistersAsyncStore object is created and
+// calls to GetRegisterValues will return a storage.ErrHeightNotIndexed,
+// since we can't disambiguate between the underlying store before bootstrapping or just simply being behind sync
+func (r *RegistersAsyncStore) InitDataAvailable(
+	indexReporter state_synchronization.IndexReporter,
+	registers storage.RegisterIndex,
+) {
+	r.init.Do(func() {
+		defer r.initialized.Store(true)
+		r.reporter = indexReporter
+		r.registerIndex = registers
+	})
+}
+
+// RegisterValues atomically gets the register values from the underlying storage.RegisterIndex
+// Expected errors:
+//   - storage.ErrHeightNotIndexed if the store is still bootstrapping or if the values at the height is not indexed yet
```

Comment:
I wonder whether these should be two errors? should the client convey/handle differently based on each case? I imagine handling bootstrapping in progress would potentially mean (not sure about use case) forwarding the request to another API that has the data, while height not indexed might just mean we can show to the user data is not available. Again not sure about the use case, but wanted to raise the question.